### PR TITLE
Generate tileset route lines and stops 

### PR DESCRIPTION
**Pull request by arindam1993**
- Adds `tippecanoe` as a git submodule dependency
- uses it to generate routeline tiles

### CODE REVIEW

File: src/generate-route-tiles.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      delete feature.properties.routes;
+    } 
+  }
+}
+
+async function generateRouteTiles(
+  zippedGtfsPath,
+  outputPath,
+) {
+  const gtfsToGeoJSON = await import('gtfs-to-geojson');
```
Comment (graue) [inline src/generate-route-tiles.js:23]: Why do you import asynchronously? I see no advantage. This is server-side code; we already have the library installed.

File: Dockerfile [hunk 2]
```diff
@@ -1,6 +1,10 @@
 FROM node:20-slim
+RUN apt-get update && apt-get -y install dumb-init gcc g++ make libsqlite3-dev zlib1g-dev
```
Comment (Andykmcc) [inline Dockerfile:2]: can you convert this to multistage build? I think it will save us a fair bit on bandwidth and speed up pulling it locally. most of our bikehopper dockerfiles are multistage builds (e.g. [bikehopper-ui](https://github.com/bikehopper/bikehopper-ui/blob/main/Dockerfile)). i can help you with it if you want, or just do it myself.

**Review by Andykmcc — APPROVED** [review]

**Comment by arindam1993** [comment]
@graue @Andykmcc This is ready for another review, its in the final form now. A bunch of changes went in since you last saw it.

File: src/generate-route-tiles.js [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    await rm(routeTilesPath, {recursive: true});
+  }
+
+  const stopTilesPath = join(outputPath, 'stop-tiles');
+  if (existsSync(stopTilesPath)) {
+    await rm(stopTilesPath, {recursive: true});
+  }
+
+  const routeLinesCommand = `tippecanoe \
+   -e ${routeTilesPath} \
```
Comment (graue) [inline src/generate-route-tiles.js:188]: This is fragile and will break if the path contains spaces or other characters with special interpretation by the shell. Use the `spawn*` functions from [child_process](https://nodejs.org/dist/latest-v20.x/docs/api/child_process.html) instead of `shell.exec` to avoid this. They take arguments as an array of strings.

File: src/get-trip-id-stop-ids-lookup.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ *    tripIdStopIdsLookup: <trip-id>: <stop-id>[],
+ *    stopIdTripIdsLookup: Map<<stop-id>, Set<trip-id>>
+ * }
+ */
+async function getStopsForTripLookup(unzippedGtfsPath){
+  const stopTimesStream = createReadStream(resolve(unzippedGtfsPath, 'stop_times.txt'), {encoding: 'utf8'});
+  const parser = stopTimesStream.pipe(parse({columns: true}));
+  const tripIdStopIdsLookup = {};
+  // Don't need to serialize the <stop-id> : <trip-ids>[] lookup, so use a Map
+  const stopIdTripIdsLookup = new Map();
```
Comment (graue) [inline src/get-trip-id-stop-ids-lookup.js:21]: You could just as well have them both be maps here, since you iterate over the object-as-map and change it below.

**Review by graue — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -4,6 +4,57 @@ Ingests a static GTFS feed and outputs useful geoJSON files. It supports excludi
...
+        "minzoom": 8,
+        "maxzoom": 14,
+        "fields": {
+            // Encoding the types of modalities at the stop
+            "bus": "Boolean",
+            "cable": "Boolean",
+            "ferry": "Boolean",
+            "rail": "Boolean",
+            "tram": "Boolean"
+            "subway": "Boolean",
```
[the remark this answers is not in the document]
Comment (graue) [reply README.md:50]: I see now that this _is_ used on the frontend: `isBus`. But the frontend only checks for "bus" and not "trolleybus," which are distinct modes in GTFS (Muni happens to not use "trolleybus"; other agencies might).

This is exactly what I was worried about in converting from the GTFS numbers to strings. Leave it as GTFS mode number, and then we can use the GTFS spec as the source of truth on all the possibilities that have to be accommodated.

[the remark this answers is not in the document]
Comment (arindam1993) [reply README.md:50]: Not much to worry. The I'm accounting for everything in the gtfs spec right here
https://github.com/bikehopper/gtfs-processor/pull/6/files#diff-6447d4336bd83e283e13fb88f8d21ef361a6095392d77f3cc16b07315d84ef26R68

Its using the exact words from the official schema as keys too. The readme was missing it because I copy-pasted the schema generated from the tileset metadata. that didn't have the other modes because of the exact reason you described.

I was originally gonna use the numbers too, but ended up doing this because it makes the `isBus` expression more readable on the client side. And those expressions are already hard to parse

File: src/get-route-id-trip-id-shape-id-lookup.js [hunk 6]
```diff
@@ -6,24 +6,30 @@ async function getRouteTripShapeLookup(unzippedGtfsPath) {
   const tripsStream = createReadStream(resolve(unzippedGtfsPath, 'trips.txt'), {encoding: 'utf8'});
   const parser = tripsStream.pipe(parse({columns: true}));
 
-  const table = {};
+  const routeTripShapeLookup = {};
+  // Using a Map for trip-id: route-id[] because this one doesn't need to be serialized to disk
```
[the remark this answers is not in the document]
Comment (arindam1993) [reply src/get-route-id-trip-id-shape-id-lookup.js:10]: Its not, I'm a firm believer in setting up data-structures for their end-to-end usecase.
Extra O(n) time and O(n) space to serialize this multi-MB dictionary back to disk isn't worth it.

File: src/get-trip-id-stop-ids-lookup.js (hunk 4 above)
Comment (arindam1993) [reply src/get-trip-id-stop-ids-lookup.js:21]: Yea, but again I don't wanna allocate a full copy of this dictionary, just to mutate one of its props.|

This talks is the root of most of my philosophy around this
http://youtube.com/watch?v=rX0ItVEVjHc&t=975s

TLDR: Good memory management makes things go fast!

File: src/generate-route-line-clipping-lookup-tables.js [hunk 7]
```diff
@@ -13,35 +14,55 @@ const { resolve } = require("path");
...
+ * }
  */
 async function generateRouteLineClippingLookupTables(
   unzippedGtfsPath,
   outputPath,
 ) {
 
   console.log('Starting build of routeline clipping tables');
-  const stopTripShapeLookup = await getRouteTripShapeLookup(unzippedGtfsPath);
+  const {routeTripShapeLookup: stopTripShapeLookup, tripRouteLookup} = await getRouteTripShapeLookup(unzippedGtfsPath);
```
[the remark this answers is not in the document]
Comment (arindam1993) [reply src/generate-route-line-clipping-lookup-tables.js]: Done, this was a mis-name left over from the previous line-clipping work.

File: src/generate-route-line-clipping-lookup-tables.js [hunk 8]
```diff
@@ -13,35 +14,55 @@ const { resolve } = require("path");
...
-    JSON.stringify(routlineLookups),
+    JSON.stringify({
+      stopTripShapeLookup,
+      shapeIdLineStringLookup,
+      tripIdStopIdsLookup,
+    }),
     'utf8',
   );
+
+  return {
```
[the remark this answers is not in the document]
Comment (arindam1993) [reply src/generate-route-line-clipping-lookup-tables.js:42]: Done!

File: src/get-trip-id-stop-ids-lookup.js (hunk 4 above)
Comment (graue) [reply src/get-trip-id-stop-ids-lookup.js:21]: Is the map big enough this matters?

File: README.md [hunk 9]
```diff
@@ -4,6 +4,62 @@ Ingests a static GTFS feed and outputs useful geoJSON files. It supports excludi
...
+```
+"vector_layers": [
+    {
+        "id": "stops",
+        "description": "",
+        "minzoom": 8,
+        "maxzoom": 14,
+        "fields": {
+            // Encoding the types of modalities at the stop
+            // Keys are taken from https://gtfs.org/documentation/schedule/reference/#routestxt
```
Comment (graue) [inline README.md:45]: I'd still prefer you used the numbers. That way, it's clearer that GTFS is the source of truth, which makes us go to the GTFS spec to see what the numbers mean, which will make us remember that, for example, we have to account for trolleybus as well as bus.

I'll accept this citation as enough of a compromise not to block on this, but it's still something that could be overlooked at the site of consumption of the data.

Maybe in the future when we combine gtfs-processor, bikehopper-web-app and bikehopper-ui into one repo, we can have a common type specification for this to mitigate the issue.

**Review by graue — APPROVED** [review]
Another comment that I would still be happier if you would address, but won't block at this point. Thanks.
